packages/@lwc/rollup-plugin/src/index.js

```diff
@@ -31,29 +31,57 @@ function isMixingJsAndTs(importerExt, importeeExt) {
     );
 }
 
+function validateUserModules(modules = []) {
+    if (modules.length) {
+        const hasNpmModuleRecords = modules.some(lwcResolver.isNpmModuleRecord);
+        if (hasNpmModuleRecords) {
+            throw new Error('Resolution of npm modules are not allowed in rollup config');
+        }
+    }
+}
+
+function validateResolvedModuleRecords(moduleRecords = []) {
+    const visited = new Set();
+    if (moduleRecords.length) {
+        moduleRecords.forEach(({ specifier, scope, version }) => {
+            const id = `${specifier}:${version}:${scope}`;
+            if (visited.has(id)) {
+                throw new Error(`Duplicated entry: ${id}`);
+            } else {
+                visited.add(id);
+            }
+        });
+    }
+}
+
 module.exports = function rollupLwcCompiler(pluginOptions = {}) {
     const { include, exclude } = pluginOptions;
     const filter = pluginUtils.createFilter(include, exclude);
     const mergedPluginOptions = Object.assign({}, DEFAULT_OPTIONS, pluginOptions);
 
-    // Closure to store the resolved modules
-    let modulePaths = {};
+    let customResolvedModules;
 
     return {
         name: 'rollup-plugin-lwc-compiler',
-
         options({ input }) {
             const { modules: userModules = [], rootDir } = mergedPluginOptions;
             const defaultModulesDir = rootDir ? path.resolve(rootDir) : path.dirname(input);
-            const modules = [...userModules, ...DEFAULT_MODULES, defaultModulesDir];
+            validateUserModules(userModules);
+            const modules = [...userModules, ...DEFAULT_MODULES, { dir: defaultModulesDir }];
             const resolvedModules = lwcResolver.resolveModules({ rootDir, modules });
```

Comment:
With this upgrade we should stop pre-caching the modules.

Reply:
There are use cases by which the iterative resolution is not sufficient.

Reply:
Let's make sure that the `resolveModule` does cover all the cases. What are the missing cases?